traffic_ops/traffic_ops_golang/test/error_context.go

```diff
@@ -0,0 +1,267 @@
+//  `code` should exist in the error context, only one mapping can exist per error context
+//  `msg` should be a plain string without special formatting
+func (ec *ErrorContext) AddMapping(code uint, msg string) error {
+	if ec.calledNewError {
+		return ec.internalError(BAD_INIT_TIMING, nil)
+	}
+	if !ec.whitelisted(code) {
+		return ec.internalError(BAD_MAP_CREATE, code)
+	}
+	if _, exists := ec.description[code]; exists {
+		return ec.internalError(BAD_DUP_MAPPING, code)
+	}
+	ec.description[code] = msg
+	return nil
+}
+
+// AddMap applies the `AddMapping` function for every element in the given map
+// the function argument will not override the current contents of the map, everything is additive
+// an error is returned if a duplicate code is added
+func (ec *ErrorContext) AddMap(mapping map[uint]string) error {
+	for code, desc := range mapping {
+		if err := ec.AddMapping(code, desc); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// all internal errors
+const (
+	BAD_ERROR_CODE  = -1 - iota // when bad error code is given in creation of new error
+	BAD_MAP_LOOKUP              // when creating an error with no error message, default message wasn't found
+	BAD_MAP_CREATE              // when bad error code is given in creation of map
+	BAD_DUP_MAPPING             // when a mapping is already made
+	BAD_FMT_STRING              // when the fmt string isn't a string
+	BAD_INIT_TIMING             // when the error context is modifed after having created an error
+)
+
+func (ec ErrorContext) internalError(code int, offender interface{}) ErrorCoder {
+
+	var err error
+	switch code {
+	case BAD_ERROR_CODE:
+		err = fmt.Errorf("error code %v not found in whitelist for \"%s\" error context", offender, ec.name)
+	case BAD_MAP_CREATE:
+		err = fmt.Errorf("when creating default error mapping, code %v wasn't found in the code whitelist for the \"%s\" error context", offender, ec.name)
+	case BAD_DUP_MAPPING:
+		err = fmt.Errorf("when creating default error mapping, code %v was already found as a mapping for the \"%s\" error context", offender, ec.name)
+	case BAD_MAP_LOOKUP:
+		err = fmt.Errorf("bad default error lookup for code %v in the \"%s\" error context", offender, ec.name)
+	case BAD_FMT_STRING:
+		err = fmt.Errorf("the leading argument \"%v\" could not be interpreted as a format string", offender)
+	case BAD_INIT_TIMING:
+		err = fmt.Errorf("tried to modify error context after creating error")
+	}
+
+	if ec.doPanic {
+		panic(err)
+	}
+	return AddErrorCode(code, err)
+}
+
+// whitelist is defined by code map membership
+func (ec ErrorContext) whitelisted(code uint) bool {
+	_, ok := ec.codes[code]
+	return ok
+}
+
+// makes sure every created error gets counted
+func (ec *ErrorContext) newError(code uint, fmtStr string, fmtArgs ...interface{}) ErrorCoder {
+	ec.codes[code]++
+	ec.createdNewError = true
+	return NewError(int(code), fmtStr, fmtArgs...)
+}
+
+// ErrorContext.NewError creates an error similar to test.NewError
```

Comment:
GoDoc, should be `// NewError creates...`

Reply:
Also, GoDoc comments should be a paragraph of complete sentences, with capitalization and punctuation. i.e.
```
// ErrorContext.NewError creates an error similar to test.NewError.
// Any error created in this manner must have a code that belongs to the error context.
// The args is interpreted as a format string with args, but I'm using a ...interface{} because
// if no args are specified a lookup will be made to find the default configured string (see `AddMapping`).
```

See https://blog.golang.org/godoc-documenting-go-code